[WIP] feat/install-xpdf-language-packages

elonzh opened the pull request:
close issue #69

kermitt2 commented:
Thank you @elonzh ! 

Installing the language support packages is necessary and it works. 

Problem, with your script, the packages are installed in a global directory, and the config file `xpdfrc` in your home directory as `.xpdfrc` - so it makes the installation not portable. I am looking now for a way to get the language packages and the config file all installed locally and built locally with the pdfalto binary, so that it can be shipped all together.

kermitt2 commented:
Dear @elonzh 

Ok it was more complicated than I thought !

xpdf is not working with relative paths at all, it expects absolute paths for the `xpdfrc` file and for the installation paths of the language resources. Of course with absolute paths, it's impossible to have something portable and to be integrated in GROBID for instance.  

For `xpdfrc`, it was not too complicated to fix - I just modified pdfalto to load the config file alongside the pdfalto executable and use that to init xpdf. So if `xpdfrc` is located with `pdfalto` executable, it will be loaded.

For the installation paths of the language packages, I don't see any other solution than patching xpdf... it's bad, but then straightforward, with the packages installed alongside the executable too with the right path created at parsing time.

From that, I've added a slighted modified version of your script in [install_deps.sh](https://github.com/kermitt2/pdfalto/blob/language_packages/install_deps.sh#L154), so that everything is install with one script. The changes are:
- we install all the language packages locally in pdfalto repo under an available `./languages/` subdirectory
- we copy a precooked `xpdfrc` at the root of the pdfalto repo with the right relative paths to the install language fonts

This way `xpdfrc` is alongside with the executable `pdfalto` and will be read, using the local resources paths if installed.

When shipping pdfalto in another application like GROBID, we simply copy "xpdfrc" and the `languages` directory together the executable `pdfalto`:

```
lopez@work:~/tmp4/pdfalto$ ll ~/grobid/grobid-home/pdfalto/lin-64/
total 36M
drwxrwxr-x  3 lopez lopez 4.0K Apr  6 01:30 ./
drwxrwxr-x  7 lopez lopez 4.0K Apr  5 22:56 ../
drwxrwxr-x 13 lopez lopez 4.0K Apr  6 01:30 languages/
-rwxrwxr-x  1 lopez lopez  36M Apr  6 04:30 pdfalto*
-rw-rw-r--  1 lopez lopez 3.4K Apr  6 01:33 xpdfrc
```

-> this is in the branch https://github.com/kermitt2/pdfalto/tree/language_packages
-> the modified `xpdf-4.00` is the one used by the updated git submodule
-> branch https://github.com/kermitt2/grobid/tree/yaml-config for the grobid version with language support packages

elonzh commented:
> Dear @elonzh
> 
> Ok it was more complicated than I thought !
> 
> xpdf is not working with relative paths at all, it expects absolute paths for the `xpdfrc` file and for the installation paths of the language resources. Of course with absolute paths, it's impossible to have something portable and to be integrated in GROBID for instance.
> 
> For `xpdfrc`, it was not too complicated to fix - I just modified pdfalto to load the config file alongside the pdfalto executable and use that to init xpdf. So if `xpdfrc` is located with `pdfalto` executable, it will be loaded.
> 
> For the installation paths of the language packages, I don't see any other solution than patching xpdf... it's bad, but then straightforward, with the packages installed alongside the executable too with the right path created at parsing time.
> 
> From that, I've added a slighted modified version of your script in [install_deps.sh](https://github.com/kermitt2/pdfalto/blob/language_packages/install_deps.sh#L154), so that everything is install with one script. The changes are:
> 
> * we install all the language packages locally in pdfalto repo under an available `./languages/` subdirectory
> * we copy a precooked `xpdfrc` at the root of the pdfalto repo with the right relative paths to the install language fonts
> 
> This way `xpdfrc` is alongside with the executable `pdfalto` and will be read, using the local resources paths if installed.
> 
> When shipping pdfalto in another application like GROBID, we simply copy "xpdfrc" and the `languages` directory together the executable `pdfalto`:
> 
> ```
> lopez@work:~/tmp4/pdfalto$ ll ~/grobid/grobid-home/pdfalto/lin-64/
> total 36M
> drwxrwxr-x  3 lopez lopez 4.0K Apr  6 01:30 ./
> drwxrwxr-x  7 lopez lopez 4.0K Apr  5 22:56 ../
> drwxrwxr-x 13 lopez lopez 4.0K Apr  6 01:30 languages/
> -rwxrwxr-x  1 lopez lopez  36M Apr  6 04:30 pdfalto*
> -rw-rw-r--  1 lopez lopez 3.4K Apr  6 01:33 xpdfrc
> ```
> 
> -> this is in the branch https://github.com/kermitt2/pdfalto/tree/language_packages
> -> the modified `xpdf-4.00` is the one used by the updated git submodule
> -> branch https://github.com/kermitt2/grobid/tree/yaml-config for the grobid version with language support packages (`lin-64` only)

If we need a portable version, maybe we can package these assets into the binary? I am not familiar with c++, but using [embedded files](https://golang.org/pkg/embed/) in Golang for portability is a common practice.

kermitt2 commented:
For docker it is indeed no problem yes! 

Embeddings in the binaries is doable, but it's more complicated and it supposes to patch much deeper xpdf which is using file intensively for this (and it will be harder and harder to update xpdf, deviating so much from the "official" one). 

Another argument agaist the binary embedding of the assets from the GROBID point of view, is that as we ship several binary versions of pdfalto for different platform, it will duplicate the resources in each binaries, while it could be share from a single resource directory when working with a relative/local install directory.

kermitt2 commented:
So I included the feature and your script in #119 ... many thanks @elonzh !
